### PR TITLE
Proposal for the FIPS crypto module loading for SGX

### PR DESCRIPTION
**Pull request by mingweishih**
Signed-off-by: Ming-Wei Shih <mishih@microsoft.com>

### CODE REVIEW

File: docs/DesignDocs/CryptoModuleLoadingSupport.md [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+[SymCrypt](https://github.com/microsoft/SymCrypt).
+
+While multiple loader implementations exist---e.g., the [dynamic linker](https://git.musl-libc.org/cgit/musl/tree/ldso/dynlink.c)
+from musl---they may require more design considerations and code changes, which are not necessary for
+supporting only the cryptographic module loading. However, such implementation would still
+be ideal for the long term if we plan to generally support shared libraries loading in OE.
+Unless there are more use cases show up, we leave this as future work and discuss possible options in
+the [Alternative Loader Implementations](#alternative-loader-implementations) section.
+
+Note that the proposal targets SGX only, which introduces modifications to the loader that
```
Comment (dthaler) [inline docs/DesignDocs/CryptoModuleLoadingSupport.md:32]: Please update the title in line 1 to say SGX then

**Review by radhikaj — APPROVED** [review]

**Review by anakrish — APPROVED** [review]
LGTM. Thanks for the focused proposal addressing FIPS crypto needs and starting points for long-term loader plans.

**Review by CodeMonkeyLeet — COMMENTED** [review]
Scope of the design change seems reasonable to me although you may want to consider expanding the discussion on the failure handling codepaths, particularly if they related to specific feature scoping decisions that need to be reflected to the developer (e.g. expected error codes/messages for unsupported relocations, test scenarios for them etc.)

On minor bug that needs to be additionally addressed is that `dl_iterate_phdr` in oelibc will be incorrect once there is more that one module in the enclave (it will only enumerate the primary enclave). It would be worth tracking as a separate issue since it's slightly orthogonal to the content of this design, and easy to overlook.

**Comment by anakrish** [comment]
Good point @CodeMonkeyLeet. Thanks for pointing out `dl_iterate_phdr`. 

Due to the `dl` prefix, it seemed as if this function belongs in `libdl`, and I'd overlooked it since we don't support `libdl`.
But good that this belongs in `libc` instead, which implies we ought to be able to support it correctly.

This ought to be the recommended way for iterating over modules within an enclave.
@mingweishih Would `dl_iterate_phdr` provide access to all the information for FIPS CMAC computation?

File: docs/DesignDocs/CryptoModuleLoadingSupport.md [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+certain cryptographic operations through a FIPS-validated module. The module usually
+takes the form of a shared library. Once loaded, the module performs a series of self-tests
+(also known as power-up self-tests, or POST), which validates both the integrity of the module
+and the correctness of the cryptographic operations.
+
+Objective
+---------
+
+This proposal aims to add the support of loading a FIPS-validated module in the OE
+SDK in a timely manner, which allows an enclave to use the module and conforms the FIPS requirement.
```
Comment (dthaler) [inline docs/DesignDocs/CryptoModuleLoadingSupport.md:19]:
```suggestion
SDK in a timely manner, which allows an enclave to use the module and conform to the FIPS requirement.
```

File: docs/DesignDocs/CryptoModuleLoadingSupport.md [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+
+  ```
+  clang -o enclave enc.o libsymcrypt.so -loeenclave -loelibc -oecore
+  ```
+
+  Note that the module is required to be put under the same directory as the enclave binary.
+  Moreover, both the `rpath` and `runpath` linker options *should not* be used.
+  See [Module Lookup](#module-lookup) for more detail.
+  The OE Make settings (i.e., `pkgconfig`) does not use these options by default and
+  CMake settings explicitly opts out these options by setting the `CMAKE_SKIP_RPATH` to `TRUE`.
```
Comment (dthaler) [inline docs/DesignDocs/CryptoModuleLoadingSupport.md:58]:
```suggestion
  CMake settings explicitly opts out of these options by setting the `CMAKE_SKIP_RPATH` to `TRUE`.
```

File: docs/DesignDocs/CryptoModuleLoadingSupport.md [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+
+  Again, this assumes that the `enclave/libsymcrypto.so` exists.
+  For cases such as the shared library is not found in the enclave directory or the library
+  is not supported (see the [Specification](#specification) for more detail), the loader
+  aborts the execution and prints proper messages.
+
+Specification
+------
+
+- The design supports a dynamically linked enclave binary against a *single* share library.
```
Comment (dthaler) [inline docs/DesignDocs/CryptoModuleLoadingSupport.md:85]:
```suggestion
- The design supports a dynamically linked enclave binary against a *single* shared library.
```

File: docs/DesignDocs/CryptoModuleLoadingSupport.md [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+
+  For a shared library with unsupported properties, the loader will explicitly error out.
+- The design retains the existing behavior if the enclave is statically linked.
+- The design targets SGX only.
+
+Design
+------
+
+Instead of supporting dynamic loading or `dlopen`, which requires Enclave Dynamic Memory Management
+(EDMM) feature in SGX2 that is not available yet, the idea of the design is supporting static loading
```
Comment (dthaler) [inline docs/DesignDocs/CryptoModuleLoadingSupport.md:102]:
```suggestion
(EDMM) in SGX2 that is not available yet, the idea of the design is supporting static loading
```

**Review by dthaler — APPROVED** [review]

**Comment by mingweishih** [comment]
> Good point @CodeMonkeyLeet. Thanks for pointing out `dl_iterate_phdr`.
> 
> Due to the `dl` prefix, it seemed as if this function belongs in `libdl`, and I'd overlooked it since we don't support `libdl`.
> But good that this belongs in `libc` instead, which implies we ought to be able to support it correctly.
> 
> This ought to be the recommended way for iterating over modules within an enclave.
> @mingweishih Would `dl_iterate_phdr` provide access to all the information for FIPS CMAC computation?

`dl_iterate_phdr` will be a callback to the module, so it may not be used to perform integrity test. However, I'll have another issue to track the problem or address it as part of actual implementation to the proposal.

**Pull request merged**